File: packages/ts/server/src/lifecycle.ts
import type { ServerPlugin } from "@drakkar.software/starfish-protocol"

export interface GracefulShutdownOptions {
  /** Called during shutdown to perform cleanup (e.g. close DB connections). */
  onShutdown?: () => Promise<void>
  /** Maximum time in ms to wait for cleanup before forcing exit. Default: 10000. */
  timeoutMs?: number
  /** Server plugins. Each plugin's `shutdown` hook runs during shutdown (e.g.
   *  `starfish-queuing` closes its queue, `starfish-replica` stops its timers). */
  plugins?: ServerPlugin[]
  /** Signals to listen for. Default: ["SIGTERM", "SIGINT"]. */
  signals?: string[]
}

export interface ShutdownHandle {
  /** Manually trigger shutdown (e.g. from tests). */
  shutdown: () => Promise<void>
  /** Remove signal listeners. */
  unregister: () => void
}

export function createGracefulShutdown(opts: GracefulShutdownOptions = {}): ShutdownHandle {
  const timeoutMs = opts.timeoutMs ?? 10_000
  const signals = opts.signals ?? ["SIGTERM", "SIGINT"]
  let shuttingDown = false

  const shutdown = async (): Promise<void> => {
    if (shuttingDown) return
    shuttingDown = true

    const timer = setTimeout(() => {
      console.error("[Starfish] Graceful shutdown timed out, forcing exit")
      if (typeof process !== "undefined") process.exit(1)
    }, timeoutMs)

    try {
      if (opts.plugins) {
        for (const p of opts.plugins) {
          if (p.shutdown) await p.shutdown()
        }
      }
      if (opts.onShutdown) {
        await opts.onShutdown()
      }
    } catch (e) {
      console.error("[Starfish] Error during graceful shutdown:", e)
    } finally {
      clearTimeout(timer)
    }
  }

  const handler = () => {
    shutdown().then(() => {
      if (typeof process !== "undefined") process.exit(0)
    })
  }

  for (const sig of signals) {
    if (typeof process !== "undefined" && process.on) {
      process.on(sig, handler)
    }
  }

  const unregister = (): void => {
    for (const sig of signals) {
      if (typeof process !== "undefined" && process.removeListener) {
        process.removeListener(sig, handler)
      }
    }
  }

  return { shutdown, unregister }
}

File: packages/python/server/starfish_server/lifecycle.py
"""Graceful shutdown utilities for Starfish servers."""


import asyncio
import inspect
import logging
import signal
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any, Callable, Awaitable

if TYPE_CHECKING:
    from starfish_protocol.plugins import ServerPlugin

logger = logging.getLogger(__name__)


@dataclass
class GracefulShutdownOptions:
    """Configuration for graceful shutdown."""
    on_shutdown: Callable[[], Awaitable[None]] | None = None
    timeout_s: float = 10.0
    plugins: "list[ServerPlugin] | None" = None
    """Server plugins. Each plugin's ``shutdown`` hook runs during shutdown
    (e.g. ``starfish-queuing`` closes its queue, ``starfish-replica`` stops its timers)."""
    signals: list[signal.Signals] = field(
        default_factory=lambda: [signal.SIGTERM, signal.SIGINT],
    )


class GracefulShutdown:
    """Manages graceful shutdown of Starfish server resources.

    Usage::

        shutdown = GracefulShutdown(GracefulShutdownOptions(
            plugins=[replica_plugin, queuing_plugin],
        ))
        shutdown.register()
        # ... run server ...
        shutdown.unregister()
    """

    def __init__(self, opts: GracefulShutdownOptions | None = None) -> None:
        self._opts = opts or GracefulShutdownOptions()
        self._shutting_down = False
        self._handlers: list[tuple[signal.Signals, Any]] = []

    async def shutdown(self) -> None:
        """Perform graceful shutdown of all managed resources."""
        if self._shutting_down:
            return
        self._shutting_down = True
        logger.info("Starfish graceful shutdown initiated")

        try:
            async with asyncio.timeout(self._opts.timeout_s):
                for plugin in self._opts.plugins or []:
                    if plugin.shutdown is not None:
                        result = plugin.shutdown()
                        if inspect.isawaitable(result):
                            await result
                if self._opts.on_shutdown:
                    await self._opts.on_shutdown()
        except asyncio.TimeoutError:
            logger.error("Starfish graceful shutdown timed out after %.1fs", self._opts.timeout_s)
        except Exception:
            logger.exception("Error during Starfish graceful shutdown")

        logger.info("Starfish graceful shutdown complete")

    def register(self, loop: asyncio.AbstractEventLoop | None = None) -> None:
        """Register signal handlers for graceful shutdown."""
        _loop = loop or asyncio.get_event_loop()
        for sig in self._opts.signals:
            _loop.add_signal_handler(sig, lambda s=sig: asyncio.ensure_future(self.shutdown()))
            self._handlers.append((sig, None))

    def unregister(self, loop: asyncio.AbstractEventLoop | None = None) -> None:
        """Remove signal handlers."""
        _loop = loop or asyncio.get_event_loop()
        for sig, _ in self._handlers:
            try:
                _loop.remove_signal_handler(sig)
            except (ValueError, OSError):
                pass
        self._handlers.clear()
